casper/src/test/scala/coop/rchain/casper/batch2/BlockReceiverEffectsSpec.scala

```diff
@@ -0,0 +1,243 @@
+        _     <- incomingQueue.enqueue1(block)
+      } yield {
+        bs.put(*) wasNever called
+        bs.contains(*) wasNever called
+        br.ackReceived(*) wasNever called
+        dagStorageWasNotModified(bds)
+        outStream should notEmit
+      }
+  }
+
+  it should "discard known block" in withEnv("root") {
+    case (incomingQueue, _, outStream, bs, br, bds) =>
+      for {
+        block <- addBlock(bs)
+        _     <- incomingQueue.enqueue1(block)
+      } yield {
+        bs.put(Seq((block.blockHash, block))) wasCalled once
+        bs.contains(*) wasNever called
+        br.ackReceived(*) wasNever called
+        dagStorageWasNotModified(bds)
+        outStream should notEmit
+      }
+  }
+
+  it should "pass to output blocks with resolved dependencies" in withEnv("root") {
+    case (incomingQueue, validatedQueue, outStream, bs, br, bds) =>
+      for {
+        // Received a parent with an empty list of justifications and its child
+        a1 <- makeBlock()
+        a2 <- makeBlock(List(a1.blockHash))
+
+        // Put the parent and child in the input queue
+        _ <- incomingQueue.enqueue1(a2)
+        _ <- incomingQueue.enqueue1(a1)
+
+        // Dependencies of the child (its parent) have not yet been resolved,
+        // so only the parent goes to the output queue, since it has no dependencies
+        a1InOutQueue <- outStream.take(1).compile.toList.map(_.head)
+
+        // A1 is now validated (e.g. in BlockProcessor)
+        _ <- validatedQueue.enqueue1(a1)
+
+        // All dependencies of child A2 are resolved, so it also goes to the output queue
+        a2InOutQueue <- outStream.take(1).compile.toList.map(_.head)
+      } yield {
+        val bsPutCaptor = ArgCaptor[Seq[(BlockHash, BlockMessage)]]
+        bs.put(bsPutCaptor) wasCalled twice
+        bsPutCaptor.values should contain allOf (Seq((a1.blockHash, a1)), Seq((a2.blockHash, a2)))
+
+        val bsContainsCaptor = ArgCaptor[Seq[BlockHash]]
+        bs.contains(bsContainsCaptor) wasCalled 3.times
+        bsContainsCaptor.values should contain allOf (Seq(a1.blockHash), Seq(a2.blockHash))
+
+        val brAckReceivedCaptor = ArgCaptor[BlockHash]
+        br.ackReceived(brAckReceivedCaptor) wasCalled twice
+        brAckReceivedCaptor.values should contain allOf (a1.blockHash, a2.blockHash)
+
+        dagStorageWasNotModified(bds)
+        a1InOutQueue shouldBe a1.blockHash
+        a2InOutQueue shouldBe a2.blockHash
+      }
+  }
+
+  private def blockDagStorageMock[F[_]: Applicative](): BlockDagStorage[F] = {
+    val emptyDag = DagRepresentation(Set(), Map(), SortedMap(), DagMessageState(), Map())
+    mock[BlockDagStorage[F]].getRepresentation returnsF emptyDag
+  }
+
+  private def blockRetrieverMock[F[_]: Applicative](): BlockRetriever[F] =
+    mock[BlockRetriever[F]].ackReceived(*) returns ().pure[F]
+
+  private def blockStoreMock[F[_]: Sync: Applicative](): BlockStore[F] = {
```

Comment:
`Sync` extends `Applicative` already. Even in cats effect 3.
https://github.com/typelevel/cats-effect/blob/series/3.x/kernel/shared/src/main/scala/cats/effect/kernel/Sync.scala#L26

---

Functions returning F value and without input can be defined without empty parenthesis because F value already behaves as lazy evaluated. It can be `val` instead of `def` if there is no type parameters.